# Preparing the data: Carbon dioxide emissions

We're walking through our four datasets, processing one at a time:

1. **Carbon dioxide emissions by country**
2. Income (as measured by GDP per capita) by country
3. Population by country (so we can convert CO2 emissions into per capita emissions)
4. List of territories by continent (since we want to be able to group the countries by region of the world)

Let's start by loading our data into a pandas DataFrame and take a look at the first 10 rows of content:

In [10]:
import pandas as pd

co2_all = pd.read_csv("data/co2_emissions_tonnes_per_person.csv")
co2_all.head(10)

country  1799  1800  1801  1802  1803  1804  1805   1806  \
0           Afghanistan   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
1                Angola   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
2               Albania   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
3               Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
4  United Arab Emirates   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
5             Argentina   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
6               Armenia   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
7   Antigua and Barbuda   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
8             Australia   NaN   NaN   NaN   NaN   NaN   NaN   NaN    NaN   
9               Austria   NaN   NaN   NaN   NaN   NaN   NaN   NaN  0.054   

   1807  ...    2008   2009    2010    2011   2012    2013    2014    2015  \
0   NaN  ...   0.238   0.29   0.406   0.345   0.28   0.253   0.262   0.245   
1   NaN  ...   1.230   1.24   1.250   1.350   1.28   1.640   1.220   1.180   
2   NaN  ...   1.470   1.56   1.790   1.690   1.69   1.900   1.600   1.570   
3   NaN  ...   6.120   6.12   5.870   5.920   5.90   5.830   5.970   6.070   
4   NaN  ...  20.900  18.30  18.900  23.800  23.70  24.200  20.700  21.700   
5   NaN  ...   4.420   4.57   4.610   4.600   4.56   4.560   4.640   4.600   
6   NaN  ...   1.510   1.48   1.730   1.990   1.91   1.910   1.650   1.760   
7   NaN  ...   5.880   5.96   5.750   5.800   5.73   5.700   5.840   5.900   
8   NaN  ...  18.800  18.40  18.000  17.800  17.10  16.700  16.800  17.000   
9   NaN  ...   8.100   8.60   8.300   7.950   7.97   7.490   7.700   7.700   

     2016    2017  
0   0.247   0.254  
1   1.140   1.120  
2   1.610   1.590  
3   6.270   6.120  
4  21.100  21.400  
5   4.550   4.410  
6   1.700   1.890  
7   5.890   5.880  
8  17.000  16.900  
9   7.940   7.750  

[10 rows x 220 columns]

Let's also use the `pandas` `describe()` method to summarize the contents:

In [11]:
co2_all.describe()

1799      1800      1801      1802      1803      1804      1805  \
count  5.000000  5.000000  7.000000  5.000000  6.000000  5.000000  5.000000   
mean   0.523786  0.517832  0.659911  0.570422  0.489042  0.590578  0.611746   
std    1.093672  1.085826  1.065140  1.201775  1.083367  1.224113  1.279321   
min    0.007330  0.007160  0.006980  0.006810  0.006650  0.006490  0.006330   
25%    0.042200  0.029300  0.038150  0.028300  0.050025  0.051700  0.047300   
50%    0.044200  0.043800  0.049400  0.046800  0.052050  0.053400  0.049700   
75%    0.045200  0.048900  1.018350  0.050200  0.068625  0.061300  0.055400   
max    2.480000  2.460000  2.450000  2.720000  2.700000  2.780000  2.900000   

           1806      1807      1808  ...        2008        2009        2010  \
count  6.000000  5.000000  5.000000  ...  194.000000  194.000000  194.000000   
mean   0.526380  0.599746  0.593316  ...    4.623215    4.725098    4.668183   
std    1.148889  1.252499  1.239320  ...    6.102350    6.137543    6.049405   
min    0.006180  0.006030  0.005880  ...    0.022700    0.030400    0.036600   
25%    0.043425  0.043800  0.044700  ...    0.543500    0.609000    0.622500   
50%    0.054600  0.052600  0.049200  ...    2.380000    2.420000    2.375000   
75%    0.113550  0.056300  0.056800  ...    6.142500    6.437500    6.635000   
max    2.870000  2.840000  2.810000  ...   41.500000   38.800000   39.200000   

             2011        2012        2013        2014        2015        2016  \
count  194.000000  194.000000  194.000000  194.000000  194.000000  194.000000   
mean     4.722525    4.638652    4.598566    4.496805    4.466514    4.490955   
std      6.201106    5.915767    6.024297    5.802658    5.636840    5.647153   
min      0.034200    0.041900    0.038300    0.036700    0.025400    0.024400   
25%      0.635000    0.670250    0.677000    0.672500    0.694250    0.686750   
50%      2.490000    2.460000    2.510000    2.525000    2.500000    2.535000   
75%      6.575000    6.235000    6.182500    5.895000    6.062500    6.032500   
max     42.500000   36.000000   43.100000   41.300000   38.500000   39.800000   

             2017  
count  194.000000  
mean     4.499112  
std      5.604449  
min      0.024300  
25%      0.688750  
50%      2.540000  
75%      6.015000  
max     38.000000  

[8 rows x 215 columns]

Looking at the data we learn a few things. First, the dataset contains data on CO2 starting in 1799 and going through 2017. For the plot we will be creating, we only need to plot one year of data, likely the most recent year. Since we only have data up to 2017, we'll need to use that as the common year. Going forward, we'll only concern ourselves with data from 2017.

Second, we see that many of those early years have `NaN` values, which indicates there is no data present. We can see in the "count" row of the description above that in 1799 and 1800, only 5 entries are non-empty, and this number is 194 from 2008 forward. Let's see how many countries we have in total, by using the `shape` attribute of `co2_all`:

In [12]:
co2_all.shape

(194, 220)

We have 194 rows and 194 non-empty values in the later years that are included, so we have one value for each country on this list in 2017. We'll want to use the latest data that we have to make the plot as relevant as possible, so the last thing to do is to just extract the 2017 data. We'll need the 2017 column for the data as well as the corresponding country/territory so we can combine the data later with income and population:

In [13]:
co2 = co2_all[["country", "2017"]]
co2

country    2017
0             Afghanistan   0.254
1                  Angola   1.120
2                 Albania   1.590
3                 Andorra   6.120
4    United Arab Emirates  21.400
..                    ...     ...
189                 Samoa   1.320
190                 Yemen   0.356
191          South Africa   8.100
192                Zambia   0.302
193              Zimbabwe   0.850

[194 rows x 2 columns]

One last step: since we'll eventually be merging our columns together, let's rename the '2017' column so that it is more self-explanatory and call it 'co2' using the `rename()` method:

In [14]:
co2 = co2.rename(columns={"2017": "co2"})
co2

country     co2
0             Afghanistan   0.254
1                  Angola   1.120
2                 Albania   1.590
3                 Andorra   6.120
4    United Arab Emirates  21.400
..                    ...     ...
189                 Samoa   1.320
190                 Yemen   0.356
191          South Africa   8.100
192                Zambia   0.302
193              Zimbabwe   0.850

[194 rows x 2 columns]

Lastly, let's save our data to file.

In [15]:
co2.to_csv('data/intermediate/co2.csv', index=False)

With all of this complete, let's create a function that includes each of the steps above to reproduce what we've done:

In [ ]:
import pandas as pd

# Function to process CO2 data.
def process_co2(infile,outfile):
    # Read the data
    co2_all = pd.read_csv(infile)
    
    # Select the columns that we need and rename them for ease of reference
    co2 = co2_all[["country", "2017"]]
    co2 = co2.rename(columns={"2017": "co2"})
    
    # Output the resulting data
    co2.to_csv(outfile, index=False)

# Test to make sure the results are the same as the original
raw_co2_file = "data/co2_emissions_tonnes_per_person.csv"
co2_file = 'data/intermediate/co2.csv'
process_co2(raw_co2_file,co2_file)

The CO2 emissions dataset is ready for use! Let's move on to exploring income per person.